File: sdks/go/providers/vocu/vocu.go
// Package vocu implements Vocu's incomplete wire contract directly.
package vocu

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"
	"unicode"

	"github.com/speechswitch/client/sdks/go/generated/auth"
	schema "github.com/speechswitch/client/sdks/go/generated/vocu"
	out "github.com/speechswitch/client/sdks/go/generated/vocu_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

type Options struct {
	Auth auth.Auth
	// Nil uses native HTTP without redirects, cookies or implicit credentials.
	// Overrides must honor cancellation and reject redirects and submission retries.
	Transport runtime.HTTPTransport
	BaseURL   string
	// Empty selects async for batches/splitters and stream otherwise.
	Mode string
	// Omission has no deadline. Explicit zero expires before network access.
	TimeoutMs runtime.Optional[int64]
	// Omission selects 1000 ms; explicit zero polls cooperatively without delay.
	PollIntervalMs runtime.Optional[int64]
	// Zero selects 4 MiB. This never bounds or buffers native audio.
	MaxMetadataBytes int
	// Additional exact trusted HTTP(S) origins. Downloads never receive API auth.
	AudioOrigins []string
}

type Error struct {
	Status                       runtime.Optional[int]
	JobID, RequestID, RetryAfter runtime.Optional[string]
}

func (e *Error) Error() string {
	if !e.Status.Present {
		return "Vocu generation failed"
	}
	return fmt.Sprintf("Vocu returned HTTP %d", e.Status.Value)
}

// Synthesize submits one request and returns at its headers. Always Close the
// returned stream. Async jobs are polled by Next; cancellation does not cancel
// accepted server jobs or their billing. There is no implicit queue timeout.
func Synthesize(ctx context.Context, request schema.TtsRequest, options Options) (runtime.Input[out.SynthesisItem], error) {
	if _, err := schema.ValidateRequest(request); err != nil {
		return nil, err
	}
	values, err := resolve(request, options.Mode)
	if err != nil {
		return nil, err
	}
	mode := options.Mode
	if mode == "" {
		if values.async {
			mode = "async"
		} else {
			mode = "stream"
		}
	}
	var apiKey, explicitKey, explicitToken runtime.Optional[string]
	if options.Auth.Vocu.Present {
		explicitKey, explicitToken = options.Auth.Vocu.Value.ApiKey, options.Auth.Vocu.Value.AccessToken
	}
	apiKey = explicitKey
	if !apiKey.Present {
		if value, present := os.LookupEnv("SPEECHSWITCH_VOCU_API_KEY"); present {
			apiKey = runtime.Some(value)
		} else if value, present := os.LookupEnv("VOCU_API_KEY"); present {
			apiKey = runtime.Some(value)
		}
	}
	token := apiKey
	if mode == "async" {
		if explicitKey.Present {
			token = explicitKey
		} else if explicitToken.Present {
			token = explicitToken
		}
		if !token.Present {
			if value, present := os.LookupEnv("SPEECHSWITCH_VOCU_ACCESS_TOKEN"); present {
				token = runtime.Some(value)
			} else if value, present := os.LookupEnv("VOCU_ACCESS_TOKEN"); present {
				token = runtime.Some(value)
			}
		}
	}
	if !token.Present || token.Value == "" {
		if mode == "async" {
			return nil, errors.New("Missing auth.vocu.apiKey or auth.vocu.accessToken configuration")
		}
		return nil, errors.New("Missing auth.vocu.apiKey configuration")
	}
	for _, r := range token.Value {
		if r < 33 || r > 126 {
			return nil, errors.New("Vocu credential must contain only visible ASCII characters")
		}
	}
	baseAddress := options.BaseURL
	if baseAddress == "" {
		baseAddress = "https://v1.vocu.ai"
	}
	base, _, err := parseURL(baseAddress)
	if err != nil {
		return nil, err
	}
	if base.RawQuery != "" || base.ForceQuery {
		return nil, errors.New("Vocu base URL must not contain a query")
	}
	poll := int64(1000)
	if options.PollIntervalMs.Present {
		poll = options.PollIntervalMs.Value
	}
	if poll < 0 || poll > 2147483647 || options.TimeoutMs.Present && (options.TimeoutMs.Value < 0 || options.TimeoutMs.Value > 2147483647) {
		return nil, errors.New("Vocu polling interval and timeout must be integers between 0 and 2147483647")
	}
	limit := options.MaxMetadataBytes
	if limit == 0 {
		limit = 4 * 1024 * 1024
	}
	if limit < 0 || uint64(limit) > 9007199254740991 {
		return nil, errors.New("Vocu MaxMetadataBytes must be a positive safe integer")
	}
	origins := map[string]bool{"https://storage.vocu.ai": true, "https://storage.vocu.studio": true, "https://v1.vocu.ai": true, "https://v1.vocu.studio": true}
	for _, address := range options.AudioOrigins {
		_, origin, err := parseURL(address)
		if err != nil {
			return nil, err
		}
		if address != origin {
			return nil, errors.New("Vocu AudioOrigins must contain HTTP(S) origins only")
		}
		origins[origin] = true
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if options.TimeoutMs.Present && options.TimeoutMs.Value == 0 {
		return nil, context.DeadlineExceeded
	}
	var cancel context.CancelFunc
	if options.TimeoutMs.Present {
		ctx, cancel = context.WithTimeout(ctx, time.Duration(options.TimeoutMs.Value)*time.Millisecond)
	} else {
		ctx, cancel = context.WithCancel(ctx)
	}
	transport := options.Transport
	if transport == nil {
		transport = &http.Client{CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}
	}
	s := &stream{ctx: ctx, cancel: cancel, transport: transport, base: base, endpoint: strings.TrimRight(base.String(), "/") + "/api/tts/", token: token.Value, mode: mode, limit: limit, poll: time.Duration(poll) * time.Millisecond, origins: origins}
	path := "simple-generate"
	if mode == "async" {
		path = "generate"
	}
	response, err := s.send(s.endpoint+path, values.payload, true)
	if err != nil {
		s.Close()
		return nil, err
	}
	s.headers = response.Header
	return s, nil
}

func parseURL(address string) (*url.URL, string, error) {
	invalid := errors.New("Vocu URLs must be HTTP(S) without credentials, fragments or invalid escapes")
	for _, r := range address {
		if unicode.IsSpace(r) || r < 32 || r == 127 || r == '\\' {
			return nil, "", invalid
		}
	}
	u, err := url.Parse(address)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") || u.Hostname() == "" || u.User != nil || u.Opaque != "" || strings.Contains(address, "#") {
		return nil, "", invalid
	}
	host := strings.ToLower(u.Hostname())
	if strings.Contains(host, ":") {
		host = "[" + host + "]"
	}
	if port := u.Port(); port != "" {
		number, err := strconv.Atoi(port)
		if err != nil || number < 0 || number > 65535 {
			return nil, "", invalid
		}
		if !(u.Scheme == "http" && number == 80 || u.Scheme == "https" && number == 443) {
			host += ":" + strconv.Itoa(number)
		}
	}
	return u, u.Scheme + "://" + host, nil
}

type stream struct {
	ctx                          context.Context
	cancel                       context.CancelFunc
	transport                    runtime.HTTPTransport
	base                         *url.URL
	endpoint, token, mode        string
	limit                        int
	poll                         time.Duration
	origins                      map[string]bool
	headers                      http.Header
	metadata                     runtime.JsonObject
	jobID, requestID             runtime.Optional[string]
	mutex, bodyMutex             sync.Mutex
	body                         runtime.Input[[]byte]
	closed                       atomic.Bool
	once                         sync.Once
	closeError                   error
	prepared, received, terminal bool
}

func (s *stream) closeBody() error {
	s.bodyMutex.Lock()
	body := s.body
	s.body = nil
	s.bodyMutex.Unlock()
	if body != nil {
		return body.Close()
	}
	return nil
}
func (s *stream) Close() error {
	s.closed.Store(true)
	s.cancel()
	s.once.Do(func() { s.closeError = s.closeBody() })
	return s.closeError
}
func (s *stream) send(address string, payload map[string]any, authenticated bool) (*runtime.HTTPResponse, error) {
	if err := s.ctx.Err(); err != nil {
		return nil, err
	}
	method := "GET"
	var data []byte
	var err error
	if payload != nil {
		method = "POST"
		data, err = json.Marshal(payload)
		if err != nil {
			return nil, err
		}
	}
	request, err := http.NewRequestWithContext(s.ctx, method, address, bytes.NewReader(data))
	if err != nil {
		return nil, errors.New("Invalid Vocu request URL")
	}
	if authenticated {
		request.Header.Set("Authorization", "Bearer "+s.token)
	}
	if payload != nil {
		request.Header.Set("Content-Type", "application/json")
	}
	response, err := runtime.OpenResponse(request, s.transport)
	if err != nil {
		return nil, err
	}
	if err := s.ctx.Err(); err != nil {
		response.Body.Close()
		return nil, err
	}
	headers := http.Header{}
	for key, values := range response.Header {
		for _, value := range values {
			headers.Add(key, value)
		}
	}
	response.Header = headers
	trace, retry := runtime.Optional[string]{}, runtime.Optional[string]{}
	if values, ok := headers["X-Vocu-App-Request-Id"]; ok && len(values) > 0 {
		trace = runtime.Some(values[0])
	}
	if values, ok := headers["Retry-After"]; ok && len(values) > 0 {
		retry = runtime.Some(values[0])
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		response.Body.Close()
		return nil, &Error{Status: runtime.Some(response.StatusCode), JobID: s.jobID, RequestID: trace, RetryAfter: retry}
	}
	if authenticated && trace.Present {
		s.requestID = trace
	}
	s.bodyMutex.Lock()
	if s.closed.Load() {
		s.bodyMutex.Unlock()
		response.Body.Close()
		return nil, context.Canceled
	}
	s.body = response.Body
	s.bodyMutex.Unlock()
	return response, nil
}

func (s *stream) Next(ctx context.Context) (out.SynthesisItem, error) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal || s.closed.Load() {
		return nil, io.EOF
	}
	stop := context.AfterFunc(ctx, func() { s.cancel(); s.closeBody() })
	defer stop()
	item, err := s.next(ctx)
	if cause := ctx.Err(); cause != nil {
		err = cause
	} else if cause := s.ctx.Err(); cause != nil {
		err = cause
	}
	if err != nil {
		s.terminal = true
		s.Close()
		return nil, err
	}
	if _, done := item.(out.SynthesisItemAsDone); done {
		s.terminal = true
		if err := s.Close(); err != nil {
			return nil, err
		}
	}
	return item, nil
}
func (s *stream) next(ctx context.Context) (out.SynthesisItem, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := s.ctx.Err(); err != nil {
		return nil, err
	}
	if !s.prepared {
		if err := s.prepare(ctx); err != nil {
			return nil, err
		}
		s.prepared = true
	}
	s.bodyMutex.Lock()
	body := s.body
	s.bodyMutex.Unlock()
	if body == nil {
		return nil, context.Canceled
	}
	audio, err := body.Next(ctx)
	if err != nil && err != io.EOF {
		return nil, err
	}
	if err == nil {
		s.received = true
		return out.SynthesisItemAsBytes{Value: audio}, nil
	}
	if !s.received {
		return nil, errors.New("Vocu returned no audio")
	}
	if err := s.closeBody(); err != nil {
		return nil, err
	}
	var completion out.VocuDoneEventCompletion = out.VocuDoneEventCompletionAsTransport{}
	if s.mode == "async" {
		completion = out.VocuDoneEventCompletionAsGenerated{}
	}
	done := out.VocuDoneEvent{Completion: completion, RequestId: s.requestID}
	if s.metadata != nil {
		var metadata out.VocuDoneEventMetadata = out.VocuDoneEventMetadataAsRecord{Value: s.metadata}
		done.Metadata = runtime.Some(metadata)
	}
	return out.SynthesisItemAsDone{Value: done}, nil
}

var jobIDPattern = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

func (s *stream) prepare(ctx context.Context) error {
	if s.mode == "stream" {
		if values, ok := s.headers["X-Reecho-Response-Data"]; ok && len(values) > 0 {
			if len(values[0]) > s.limit {
				return errors.New("Vocu metadata exceeds MaxMetadataBytes")
			}
			value, err := decodeObject([]byte(values[0]))
			if err != nil {
				return err
			}
			s.metadata = value
		}
	} else {
		value, err := s.readMetadata(ctx)
		if err != nil {
			return err
		}
		s.metadata = value
		if s.mode == "async" {
			id, ok := value["id"].(runtime.JsonString)
			if !ok || !jobIDPattern.MatchString(string(id)) {
				return errors.New("Invalid Vocu job ID")
			}
			s.jobID = runtime.Some(string(id))
			for {
				if value["id"] != id {
					return errors.New("Vocu returned a different job ID while polling")
				}
				status, ok := value["status"].(runtime.JsonString)
				if !ok {
					return errors.New("Invalid Vocu job status")
				}
				if status == "failed" {
					return &Error{JobID: s.jobID, RequestID: s.requestID}
				}
				if status == "generated" {
					break
				}
				if status != "pending" && status != "processing" {
					return errors.New("Invalid Vocu job status")
				}
				timer := time.NewTimer(s.poll)
				select {
				case <-timer.C:
				case <-s.ctx.Done():
					timer.Stop()
					return s.ctx.Err()
				}
				if _, err := s.send(s.endpoint+"generate/"+string(id), nil, true); err != nil {
					return err
				}
				value, err = s.readMetadata(ctx)
				if err != nil {
					return err
				}
			}
			s.metadata = value
		}
		var address runtime.JsonValue
		if s.mode == "async" {
			native, ok := s.metadata["metadata"].(runtime.JsonObject)
			if !ok {
				return errors.New("Invalid Vocu response object")
			}
			address = native["audio"]
		} else {
			address = s.metadata["streamUrl"]
			if _, isNull := address.(runtime.JsonNull); address == nil || isNull {
				address = s.metadata["audio"]
			}
		}
		location, ok := address.(runtime.JsonString)
		if !ok || strings.TrimSpace(string(location)) == "" {
			return errors.New("Vocu returned no audio URL")
		}
		for _, r := range location {
			if unicode.IsSpace(r) || r < 32 || r == 127 || r == '\\' {
				return errors.New("Vocu returned an untrusted audio URL")
			}
		}
		relative, err := url.Parse(string(location))
		if err != nil || strings.Contains(string(location), "#") {
			return errors.New("Vocu returned an untrusted audio URL")
		}
		addressURL, origin, err := parseURL(s.base.ResolveReference(relative).String())
		if err != nil || !s.origins[origin] {
			return errors.New("Vocu returned an untrusted audio URL")
		}
		response, err := s.send(addressURL.String(), nil, false)
		if err != nil {
			return err
		}
		s.headers = response.Header
	}
	media := strings.ToLower(strings.TrimSpace(strings.SplitN(s.headers.Get("Content-Type"), ";", 2)[0]))
	if media != "" && media != "audio/mpeg" && media != "application/octet-stream" {
		return errors.New("Vocu returned an unexpected audio content type")
	}
	return nil
}

func (s *stream) readMetadata(ctx context.Context) (runtime.JsonObject, error) {
	s.bodyMutex.Lock()
	body := s.body
	s.bodyMutex.Unlock()
	if body == nil {
		return nil, context.Canceled
	}
	var data []byte
	for {
		chunk, err := body.Next(ctx)
		if err != nil && err != io.EOF {
			return nil, err
		}
		if len(chunk) > s.limit-len(data) {
			return nil, errors.New("Vocu metadata exceeds MaxMetadataBytes")
		}
		data = append(data, chunk...)
		if err == io.EOF {
			break
		}
	}
	value, err := decodeObject(bytes.TrimPrefix(data, []byte{0xef, 0xbb, 0xbf}))
	closeErr := s.closeBody()
	if err != nil {
		return nil, err
	}
	if value["status"] != runtime.JsonNumber(200) {
		return nil, errors.New("Invalid Vocu response status")
	}
	result, ok := value["data"].(runtime.JsonObject)
	if !ok {
		return nil, errors.New("Invalid Vocu response object")
	}
	return result, closeErr
}

func decodeObject(data []byte) (runtime.JsonObject, error) {
	if !runtime.ValidUTF8JSON(data) {
		return nil, errors.New("Invalid Vocu metadata")
	}
	var value any
	if json.Unmarshal(data, &value) != nil {
		return nil, errors.New("Invalid Vocu metadata")
	}
	object, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("Invalid Vocu response object")
	}
	result := runtime.JsonObject{}
	for key, value := range object {
		result[key] = jsonValue(value)
	}
	return result, nil
}
func jsonValue(value any) runtime.JsonValue {
	// encoding/json has already rejected invalid syntax and nonfinite numbers.
	switch v := value.(type) {
	case nil:
		return runtime.JsonNull{}
	case bool:
		return runtime.JsonBool(v)
	case float64:
		return runtime.JsonNumber(v)
	case string:
		return runtime.JsonString(v)
	case []any:
		result := make(runtime.JsonArray, len(v))
		for i, child := range v {
			result[i] = jsonValue(child)
		}
		return result
	case map[string]any:
		result := runtime.JsonObject{}
		for key, child := range v {
			result[key] = jsonValue(child)
		}
		return result
	default:
		panic("unreachable decoded JSON type")
	}
}

File: sdks/python/speechswitch/providers/vocu.py
"""Vocu's native HTTP protocol, with TypeScript-generated public contracts."""

import asyncio
import json
import os
import re
from collections.abc import AsyncGenerator, AsyncIterator, Mapping, Sequence
from contextlib import aclosing, asynccontextmanager
from dataclasses import dataclass
from typing import Literal, cast
from urllib.parse import urljoin, urlsplit, urlunsplit

from speechswitch.generated.auth import Auth
from speechswitch.generated.vocu import TtsRequest
from speechswitch.generated.vocu_output import SynthesisItem, VocuDoneEvent
from speechswitch.generated.validators.vocu import validate_request
from speechswitch.http import AudioStream, HttpRequest, HttpTransport
from speechswitch.json import JsonValue
from speechswitch.validation import is_json_value, is_mapping


class VocuError(Exception):
    def __init__(self, status: int | None, job_id: str | None, request_id: str | None, retry_after: str | None) -> None:
        self.status, self.job_id, self.request_id, self.retry_after = status, job_id, request_id, retry_after
        super().__init__("Vocu generation failed" if status is None else f"Vocu returned HTTP {status}")


def _wire_binding(request: Mapping[str, object]) -> dict[str, object]:
    # Only called after generated validation. Bindings deliberately do not receive
    # utterance defaults: omission retains the native splitter's inherited state.
    wire: dict[str, object] = {}
    for source, target in (("voice", "voiceId"), ("voice_style", "promptId"), ("vivid_expression", "vivid"),
                           ("long_text_mode", "infinite_mode"), ("audio_processing_profile", "post_processing")):
        if source in request:
            wire[target] = request[source]
    if "delivery_mode" in request:
        wire["preset"] = "balance" if request["delivery_mode"] == "balanced" else request["delivery_mode"]
    if "language" in request:
        language = request["language"]
        wire["language"] = "en-us" if language == "en-US" else "fr-fr" if language == "fr-FR" else language
    if "speed" in request:
        wire["speechRate"] = 1 / cast(float, request["speed"])
    if "random_seed" in request:
        wire["seed"] = int(cast(float, request["random_seed"]))
    if "emotion_source" in request:
        wire["break_clone"] = request["emotion_source"] == "text"
    if "emotion_blend" in request:
        blend = cast(Mapping[str, float], request["emotion_blend"])
        wire["emo_switch"] = [int(blend.get(key, 0)) for key in ("anger", "happiness", "neutral", "sadness", "contextual")]
    if "input_type" in request:
        wire["instruct_mode"] = request["input_type"] == "markup"
        if "reference_emphasis" in request:
            wire["reference_mode"] = request["reference_emphasis"]
    return wire


def _wire_speech(request: Mapping[str, object]) -> dict[str, object]:
    return {**_wire_binding({"voice_style": "default", "delivery_mode": "balanced", "language": "auto",
                            "vivid_expression": False, "speed": 1, "random_seed": -1, **request}), "text": request["text"]}


def _wire_splitter(splitter: Mapping[str, object]) -> dict[str, object]:
    if "id" in splitter:
        return {"splitterId": splitter["id"]}
    native: dict[str, object] = {}
    placeholders = cast(Sequence[Mapping[str, object]], splitter.get("placeholders", ()))
    for index in range(len(placeholders)):
        rule = placeholders[index]
        marker = cast(str, rule["marker"])
        if marker in ("splitterMarks", "lookupTable", "fallbackConfig") or marker in native:
            raise TypeError("Vocu splitter markers must be unique and cannot use reserved protocol keys")
        native[marker] = _wire_binding(rule)
    if "brackets" in splitter:
        brackets = cast(Sequence[Mapping[str, str]], splitter["brackets"])
        native["splitterMarks"] = [brackets[index]["open"] + brackets[index]["close"] for index in range(len(brackets))]
    if "fallback" in splitter:
        native["fallbackConfig"] = _wire_binding(cast(Mapping[str, object], splitter["fallback"]))
    if "lookup" in splitter:
        rules = cast(Sequence[Mapping[str, object]], splitter["lookup"])
        lookup: dict[str, object] = {}
        for index in range(len(rules)):
            rule = rules[index]
            tags = cast(Sequence[str | Sequence[str]], rule["tags"])
            native_tags: list[str | list[str]] = []
            for tag_index in range(len(tags)):
                tag = tags[tag_index]
                native_tags.append(tag if isinstance(tag, str) else [tag[item] for item in range(len(tag))])
            lookup[f"entry{index}"] = {**_wire_binding(rule), "tags": native_tags}
        native["lookupTable"] = lookup
    return {"splitter": native}


def _object(value: object) -> Mapping[str, JsonValue]:
    if not is_mapping(value) or not is_json_value(value):
        raise TypeError("Invalid Vocu response object")
    return cast(Mapping[str, JsonValue], value)


def _metadata(text: str) -> Mapping[str, JsonValue]:
    try:
        value: object = json.loads(text)
    except (ValueError, RecursionError):
        raise TypeError("Invalid Vocu metadata") from None
    return _object(value)


def _url(address: str) -> tuple[str, str]:
    try:
        url = urlsplit(address)
        if url.scheme not in ("http", "https") or not url.hostname or url.username is not None or url.password is not None or "#" in address or any(c.isspace() or ord(c) < 32 or ord(c) == 127 or c == "\\" for c in address) or re.search(r"%(?![0-9a-fA-F]{2})", address):
            raise ValueError
        port = url.port
        host = f"[{url.hostname}]" if ":" in url.hostname else url.hostname.encode("idna").decode("ascii")
        origin = f"{url.scheme}://{host}"
        if port is not None and port != (443 if url.scheme == "https" else 80):
            origin += f":{port}"
        return urlunsplit((url.scheme, origin.split("://", 1)[1], url.path, url.query, "")), origin
    except (ValueError, UnicodeError):
        raise TypeError("Vocu URLs must be HTTP(S) without credentials, fragments or invalid escapes") from None


@dataclass
class _Session:
    transport: HttpTransport
    base_url: str
    endpoint: str
    token: str
    limit: int
    poll_ms: int
    audio_origins: set[str]
    request_id: str | None = None
    job_id: str | None = None

    @asynccontextmanager
    async def send(self, url: str, payload: Mapping[str, object] | None, *, authenticated: bool) -> AsyncIterator[tuple[Mapping[str, str], AudioStream]]:
        headers = {"Authorization": f"Bearer {self.token}"} if authenticated else {}
        encoded = b""
        if payload is not None:
            headers["Content-Type"] = "application/json"
            encoded = json.dumps(payload, allow_nan=False, separators=(",", ":")).encode("utf-8")
        response = await self.transport.send(HttpRequest("GET" if payload is None else "POST", url, headers, encoded))
        body = AudioStream(response.body)
        try:
            headers = {key.lower(): value for key, value in response.headers.items()}
            if not 200 <= response.status < 300:
                raise VocuError(response.status, self.job_id, headers.get("x-vocu-app-request-id"), headers.get("retry-after"))
            if authenticated:
                self.request_id = headers.get("x-vocu-app-request-id", self.request_id)
            yield headers, body
        except BaseException:
            try:
                await body.aclose()
            except Exception:
                pass
            raise
        finally:
            await body.aclose()

    async def metadata(self, body: AudioStream) -> Mapping[str, JsonValue]:
        data = bytearray()
        async for chunk in body:
            if len(chunk) > self.limit - len(data):
                raise TypeError("Vocu metadata exceeds max_metadata_bytes")
            data.extend(chunk)
            await asyncio.sleep(0)
        try:
            result = _metadata(data.decode("utf-8-sig"))
        except UnicodeError:
            raise TypeError("Invalid Vocu metadata UTF-8") from None
        if result.get("status") != 200:
            raise TypeError("Invalid Vocu response status")
        return _object(result.get("data"))

    def audio_url(self, address: JsonValue | None) -> str:
        if not isinstance(address, str) or not address.strip():
            raise TypeError("Vocu returned no audio URL")
        # Check the original spelling before urljoin can erase controls or a fragment.
        if "#" in address or any(c.isspace() or ord(c) < 32 or ord(c) == 127 or c == "\\" for c in address):
            raise TypeError("Vocu returned an untrusted audio URL")
        url, origin = _url(urljoin(self.base_url, address))
        if origin not in self.audio_origins:
            raise TypeError("Vocu returned an untrusted audio URL")
        return url


async def _audio(headers: Mapping[str, str], body: AudioStream) -> AsyncGenerator[bytes]:
    media = headers.get("content-type", "").split(";", 1)[0].strip().lower()
    if media and media not in ("audio/mpeg", "application/octet-stream"):
        raise TypeError("Vocu returned an unexpected audio content type")
    received = False
    async for chunk in body:
        received = True
        yield chunk
        await asyncio.sleep(0)
    if not received:
        raise TypeError("Vocu returned no audio")


async def _items(session: _Session, payload: Mapping[str, object], mode: Literal["stream", "http", "async"]) -> AsyncGenerator[SynthesisItem]:
    result: Mapping[str, JsonValue] | None = None
    completion: Literal["transport", "generated"] = "transport"
    async with session.send(session.endpoint + ("generate" if mode == "async" else "simple-generate"), payload, authenticated=True) as (headers, body):
        if mode == "stream":
            header = headers.get("x-reecho-response-data")
            if header is not None:
                if len(header.encode("utf-8")) > session.limit:
                    raise TypeError("Vocu metadata exceeds max_metadata_bytes")
                result = _metadata(header)
            async with aclosing(_audio(headers, body)) as audio:
                async for chunk in audio:
                    yield chunk
        else:
            result = await session.metadata(body)
    if mode == "async":
        assert result is not None
        job_id = result.get("id")
        if not isinstance(job_id, str) or re.fullmatch(r"[A-Za-z0-9_-]+", job_id) is None:
            raise TypeError("Invalid Vocu job ID")
        session.job_id = job_id
        while True:
            if result.get("id") != job_id:
                raise TypeError("Vocu returned a different job ID while polling")
            status = result.get("status")
            if status == "failed":
                raise VocuError(None, job_id, session.request_id, None)
            if status == "generated":
                break
            if status not in ("pending", "processing"):
                raise TypeError("Invalid Vocu job status")
            await asyncio.sleep(session.poll_ms / 1000)
            async with session.send(session.endpoint + "generate/" + job_id, None, authenticated=True) as (_, body):
                result = await session.metadata(body)
        completion = "generated"
    if mode != "stream":
        assert result is not None
        address = _object(result.get("metadata")).get("audio") if mode == "async" else result.get("streamUrl")
        if mode == "http" and address is None:
            address = result.get("audio")
        async with session.send(session.audio_url(address), None, authenticated=False) as (headers, body):
            async with aclosing(_audio(headers, body)) as audio:
                async for chunk in audio:
                    yield chunk
    await asyncio.sleep(0)
    done: VocuDoneEvent = {"event": "done", "completion": completion}
    if result is not None:
        done = {**done, "metadata": result}
    if session.request_id is not None:
        done = {**done, "request_id": session.request_id}
    yield done


@asynccontextmanager
async def synthesize(request: TtsRequest, *, transport: HttpTransport, auth: Auth | None = None,
                     base_url: str = "https://v1.vocu.ai", mode: Literal["stream", "http", "async"] | None = None,
                     timeout_ms: int | None = None, poll_interval_ms: int = 1000,
                     max_metadata_bytes: int = 4 * 1024 * 1024, audio_origins: Sequence[str] = ()) -> AsyncIterator[AsyncIterator[SynthesisItem]]:
    """Use async with, even for early exit; the deadline covers the entire context.

    The injected transport must return at headers, honor cancellation, never
    follow redirects/retry requests, and send neither cookies nor implicit auth.
    Stopping polling does not cancel an accepted server job or its billing.
    """
    validate_request(request)
    batch, splitter = request.get("segments"), request.get("text_splitter")
    selected = mode if mode is not None else "async" if batch is not None or splitter is not None else "stream"
    if selected not in ("stream", "http", "async"):
        raise TypeError("Invalid Vocu mode")
    if (batch is not None or splitter is not None) and selected != "async":
        raise TypeError("Vocu segments and text splitting require async synthesis")
    if selected == "async" and request.get("latency_optimization") == "maximum":
        raise TypeError("Vocu async synthesis does not support flash latency optimization")
    credentials = (auth or {}).get("vocu", {})
    api_key = credentials.get("api_key")
    if api_key is None:
        api_key = os.environ.get("SPEECHSWITCH_VOCU_API_KEY", os.environ.get("VOCU_API_KEY"))
    token = credentials.get("api_key", credentials.get("access_token", api_key)) if selected == "async" else api_key
    if selected == "async" and token is None:
        token = os.environ.get("SPEECHSWITCH_VOCU_ACCESS_TOKEN", os.environ.get("VOCU_ACCESS_TOKEN"))
    if not token:
        raise TypeError("Missing auth.vocu.apiKey or auth.vocu.accessToken configuration" if selected == "async" else "Missing auth.vocu.apiKey configuration")
    if any(ord(c) < 33 or ord(c) > 126 for c in token):
        raise TypeError("Vocu credential must contain only visible ASCII characters")
    base, _ = _url(base_url)
    if "?" in base_url:
        raise TypeError("Vocu base URL must not contain a query")
    for value in (poll_interval_ms, timeout_ms if timeout_ms is not None else 0):
        if type(value) is not int or not 0 <= value <= 2147483647:
            raise TypeError("Vocu polling interval and timeout must be integers between 0 and 2147483647")
    if type(max_metadata_bytes) is not int or not 0 < max_metadata_bytes <= 9007199254740991:
        raise TypeError("Vocu max_metadata_bytes must be a positive safe integer")
    origins = {"https://storage.vocu.ai", "https://storage.vocu.studio", "https://v1.vocu.ai", "https://v1.vocu.studio"}
    for address in audio_origins:
        _, origin = _url(address)
        if origin != address:
            raise TypeError("Vocu audio_origins must contain HTTP(S) origins only")
        origins.add(origin)
    payload: dict[str, object]
    if selected == "async":
        payload = {"text": request.get("text"), **_wire_splitter(splitter)} if splitter is not None else {"contents": [{"type": "text", **_wire_speech(batch[index])} for index in range(len(batch))] if batch is not None else [{"type": "text", **_wire_speech(request)}]}
        payload["srt"] = request.get("subtitle_format") == "srt"
    else:
        payload = {**_wire_speech(request), "flash": request.get("latency_optimization") == "maximum", "srt": request.get("subtitle_format") == "srt", "stream": True, "direct_stream": selected == "stream"}
    if timeout_ms == 0:
        raise TimeoutError("Vocu synthesis deadline expired")
    session = _Session(transport, base, base.rstrip("/") + "/api/tts/", token, max_metadata_bytes, poll_interval_ms, origins)
    async with asyncio.timeout(None if timeout_ms is None else timeout_ms / 1000):
        async with aclosing(_items(session, payload, selected)) as items:
            yield items
